### PR TITLE
Strategic objective tracking improvements

### PR DESCRIPTION
**Pull request by NickAragua**
Multiple improvements to strategic objective tracking, as well as several bug fixes:

- Starting altitude for generated units is now preserved across saves
- Taking a break between adding a bot and trying to change its settings to avoid "can't connect client" errors and issues sending settings over
- Strategic objectives are now explicitly set at contract initialization, in the basic form of [Type] [At coordinates (optional)]. This enables explicit tracking of objectives and much more detailed display.
- Don't apply "global" scenario modifiers once per track to reduce "Allied Trainees" and "Overwhelming Enemy Reinforcements" spam; apply them only once instead.

### CODE REVIEW

**Review by Windchild292 — COMMENTED** [review]
Few initial comments, will finish post-D&D

File: MekHQ/src/mekhq/campaign/stratcon/StratconContractInitializer.java [hunk 1]
```diff
@@ -105,17 +104,28 @@ public static void initializeCampaignState(AtBContract contract, Campaign campai
...
-                        }
-                    break;
+                        // set up a "win X scenarios" objective
+                        StratconStrategicObjective sso = new StratconStrategicObjective();
+                        sso.setDesiredObjectiveCount(numObjects);
+                        sso.setObjectiveType(StrategicObjectiveType.AnyScenarioVictory);
+                        campaignState.getTrack(x).addStrategicObjective(sso);
+                        break;
+                }
+            }
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/stratcon/StratconContractInitializer.java:120]: Getting a warning for a missing case from CI. Does RequiredScenarioVictory need to be handled here?

File: MekHQ/src/mekhq/campaign/stratcon/StratconRulesManager.java [hunk 2]
```diff
@@ -244,6 +245,7 @@ public static void deployForceToCoords(StratconCoords coords, int forceID, Campa
 
         if (isNonAlliedFacility || spawnScenario) {
             StratconScenario scenario = setupScenario(coords, forceID, campaign, contract, track);
+            setScenarioDates(0, track, campaign, scenario); // we deploy immediately in this case, since it's an
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/stratcon/StratconRulesManager.java:248]: Missing end to the comment

File: MekHQ/src/mekhq/campaign/stratcon/StratconRulesManager.java [hunk 3]
```diff
@@ -1039,7 +1044,9 @@ public static boolean isUnitDeployedToStratCon(Unit u) {
         // this is a little inefficient, but probably there aren't too many active AtB
         // contracts at a time
         return u.getCampaign().getActiveAtBContracts().stream().
-            anyMatch(contract -> contract.getStratconCampaignState().isForceDeployedHere(u.getForceId()));
+            anyMatch(contract ->
+                contract.getStratconCampaignState() != null &&
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/stratcon/StratconRulesManager.java:1048]: Brackets

File: MekHQ/src/mekhq/campaign/stratcon/StratconRulesManager.java [hunk 4]
```diff
@@ -1039,7 +1044,9 @@ public static boolean isUnitDeployedToStratCon(Unit u) {
...
+                contract.getStratconCampaignState().isForceDeployedHere(u.getForceId()));
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/stratcon/StratconRulesManager.java:1049]: Could probably do Campaign::getActiveStratConContracts, then do the check, but that can be left for later.

File: MekHQ/src/mekhq/campaign/stratcon/StratconRulesManager.java [hunk 5]
```diff
@@ -1251,9 +1258,10 @@ public static void processScenarioCompletion(ResolveScenarioTracker rst) {
                         campaignState.updateVictoryPoints(victory ? 1 : -1);
                     }
 
-                    // this must be done before updating ownership change to avoid
-                    // updating SO counts in the wrong direction.
-                    updateStrategicObjectiveCount(victory, scenario, facility, campaignState);
+                    // this must be done before removing the scenario from the track
+                    // in case any objectives are linked to the scenario's coordinates
+                    updateStrategicObjectives(victory, scenario, track, campaignState);
+                    //updateStrategicObjectiveCount(victory, scenario, facility, campaignState);
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/stratcon/StratconRulesManager.java:1264]: Why keep this code?

File: MekHQ/src/mekhq/campaign/stratcon/StratconRulesManager.java [hunk 6]
```diff
@@ -1268,6 +1276,30 @@ public static void processScenarioCompletion(ResolveScenarioTracker rst) {
         }
     }
     
+    /**
+     * Worker function that updates strategic objectives relevant to the passed in
+     * scenario, track and campaign state. For example, "win scenario A" or "win X scenarios".
+     */
+    private static void updateStrategicObjectives(boolean victory, StratconScenario scenario, 
+            StratconTrackState track, StratconCampaignState campaignState) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/stratcon/StratconRulesManager.java:1284]: Given campaignState is unused, why keep it?

File: MekHQ/src/mekhq/campaign/stratcon/StratconRulesManager.java [hunk 7]
```diff
@@ -1268,6 +1276,30 @@ public static void processScenarioCompletion(ResolveScenarioTracker rst) {
...
+        if (victory && (specificObjective != null) && 
+                (specificObjective.getObjectiveType() == StrategicObjectiveType.SpecificScenarioVictory)) {
+            specificObjective.incrementCurrentObjectiveCount();
+        }
+        
+        // "any scenario victory" is not linked to any specific coordinates, so we have to
+        // search through the track's objectives and update those.
+        for (StratconStrategicObjective objective : track.getStrategicObjectives()) {
+            if ((objective.getObjectiveType() == StrategicObjectiveType.AnyScenarioVictory) &&
+                    victory) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/stratcon/StratconRulesManager.java:1297]: Given how short victory is maybe keep it on one line

File: MekHQ/src/mekhq/campaign/stratcon/StratconTrackState.java [hunk 8]
```diff
@@ -196,6 +201,15 @@ public boolean isGmRevealed() {
...
         this.gmRevealed = gmRevealed;
     }
+    
+    /**
+     * Convenience function that determins if there are any forces
+     * deployed to the given coordinates.
+     */
+    public boolean areAnyForceDeployedTo(StratconCoords coords) {
+        return getAssignedCoordForces().containsKey(coords) &&
+                getAssignedCoordForces().get(coords).size() > 0;
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/stratcon/StratconTrackState.java:211]: bracket

File: MekHQ/src/mekhq/gui/StratconTab.java [hunk 9]
```diff
@@ -50,6 +54,9 @@
     private JComboBox<TrackDropdownItem> cboCurrentTrack;
     private JLabel infoPanelText;
     private JLabel campaignStatusText;
+    private JLabel objectiveStatusText;
+    JScrollPane expandedObjectivePanel;
+    boolean objectivesCollapsed = true;
```
Comment (Windchild292) [inline MekHQ/src/mekhq/gui/StratconTab.java:59]: Access modifiers?

File: MekHQ/src/mekhq/gui/StratconTab.java [hunk 10]
```diff
@@ -66,7 +73,24 @@ public void initTab() {
...
+        
         campaignStatusText = new JLabel();
+        campaignStatusText.setHorizontalAlignment(SwingConstants.LEFT);
+        campaignStatusText.setVerticalAlignment(SwingConstants.TOP);
+        
+        objectiveStatusText = new JLabel();
+        objectiveStatusText.setHorizontalAlignment(SwingConstants.LEFT);
+        objectiveStatusText.setVerticalAlignment(SwingConstants.TOP);
+        objectiveStatusText.addMouseListener(new MouseAdapter() { 
+            public void mousePressed(MouseEvent me) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/gui/StratconTab.java:87]: Missing Override annotation

File: MekHQ/src/mekhq/gui/StratconTab.java [hunk 11]
```diff
@@ -160,30 +189,173 @@ private void updateCampaignState() {
...
+                }
+            }
+        }
+        
+        StringBuilder sb = new StringBuilder();
+        if (completedObjectives >= desiredObjectives) {
+            sb.append("<span color='green'>");
+        } else {
+            sb.append("<span color='red'>");
+        }
```
Comment (Windchild292) [inline MekHQ/src/mekhq/gui/StratconTab.java:266]: Ternary candidate

File: MekHQ/src/mekhq/gui/StratconTab.java [hunk 12]
```diff
@@ -160,30 +189,173 @@ private void updateCampaignState() {
...
+        if (campaignState.getContract().getCommandRights() <= Contract.COM_LIAISON) {
+            desiredObjectives++;
+            
+            if (campaignState.getVictoryPoints() > 0) {
+                completedObjectives++;
+            }
+        }
+
+        sb.append("Strategic objectives: " + completedObjectives + "/" + desiredObjectives + " completed</span>");
+        return sb.toString();
```
Comment (Windchild292) [inline MekHQ/src/mekhq/gui/StratconTab.java:278]: Can probably merge

File: MekHQ/src/mekhq/gui/StratconTab.java [hunk 13]
```diff
@@ -160,30 +189,173 @@ private void updateCampaignState() {
...
+            for (StratconStrategicObjective objective : track.getStrategicObjectives()) {
+                boolean coordsRevealed = track.getRevealedCoords().contains(objective.getObjectiveCoords());
+                boolean displayCoordinateData = objective.getObjectiveCoords() != null;
+                boolean objectiveCompleted = objective.isObjectiveCompleted(track);
+                
+                if (objectiveCompleted) {
+                    sb.append("<span color='green'>");
+                } else {
+                    sb.append("<span color='red'>");
+                }
```
Comment (Windchild292) [inline MekHQ/src/mekhq/gui/StratconTab.java:305]: ternary candidate

File: MekHQ/src/mekhq/gui/StratconTab.java [hunk 14]
```diff
@@ -160,30 +189,173 @@ private void updateCampaignState() {
...
+                        sb.append(coordsRevealed ? "E" : "e");
+                        sb.append("ngage and defeat hostile forces");
+                        break;
+                    case HostileFacilityControl:
+                        sb.append(coordsRevealed ? "C" : "c");
+                        sb.append("apture or destroy designated facility");
+                        break;
+                    case AlliedFacilityControl:
+                        sb.append(coordsRevealed ? "M" : "m");
+                        sb.append("aintain control of designated facility");
```
Comment (Windchild292) [inline MekHQ/src/mekhq/gui/StratconTab.java:322]: Chain appends

File: MekHQ/src/mekhq/gui/StratconTab.java [hunk 15]
```diff
@@ -160,30 +189,173 @@ private void updateCampaignState() {
...
+                        break;
+                    case AnyScenarioVictory:
+                        sb.append("Engage and defeat hostile forces in ")
+                        .append(objective.getCurrentObjectiveCount()).append("/")
+                        .append(objective.getDesiredObjectiveCount())
+                        .append(" scenarios on ").append(track.getDisplayableName());
```
Comment (Windchild292) [inline MekHQ/src/mekhq/gui/StratconTab.java:328]: indent

File: MekHQ/src/mekhq/gui/StratconTab.java [hunk 16]
```diff
@@ -160,30 +189,173 @@ private void updateCampaignState() {
...
+            }
+        }
+        
+        // special case text reminding player to complete required scenarios
+        if (campaignState.getContract().getCommandRights() <= Contract.COM_LIAISON) {
+            if (campaignState.getVictoryPoints() > 0) {
+                sb.append("<span color='green'>");
+            } else {
+                sb.append("<span color='red'>");
+            }
```
Comment (Windchild292) [inline MekHQ/src/mekhq/gui/StratconTab.java:352]: Ternary candidate

**Review by Windchild292 — COMMENTED** [review]
Looks okay, but with a few cleanups needed

File: MekHQ/src/mekhq/campaign/stratcon/StratconContractInitializer.java (hunk 1 above)
Comment (NickAragua) [reply MekHQ/src/mekhq/campaign/stratcon/StratconContractInitializer.java:120]: No - currently, none of the contract definitions have it. I was briefly considering having the "required" scenarios from liaison+ command be a separate objective, but shelved that idea.

Comment (Windchild292) [reply MekHQ/src/mekhq/campaign/stratcon/StratconContractInitializer.java:120]: Solid idea for the future, especially when considering the CO descriptions

File: MekHQ/src/mekhq/gui/StratconTab.java (hunk 12 above)
Comment (NickAragua) [reply MekHQ/src/mekhq/gui/StratconTab.java:278]: meaning?

Comment (Windchild292) [reply MekHQ/src/mekhq/gui/StratconTab.java:278]: sb.append.toString

Comment (NickAragua) [reply MekHQ/src/mekhq/gui/StratconTab.java:279]: No need to make that line longer, I think.

File: MekHQ/src/mekhq/gui/StratconTab.java (hunk 13 above)
Comment (NickAragua) [reply MekHQ/src/mekhq/gui/StratconTab.java:306]: Gonna skip that for now.

File: MekHQ/src/mekhq/campaign/stratcon/StratconRulesManager.java [hunk 17]
```diff
@@ -1045,7 +1046,7 @@ public static boolean isUnitDeployedToStratCon(Unit u) {
         // contracts at a time
         return u.getCampaign().getActiveAtBContracts().stream().
             anyMatch(contract ->
-                contract.getStratconCampaignState() != null &&
+                (contract.getStratconCampaignState()) != null &&
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/stratcon/StratconRulesManager.java:1049]: Uhm... wrong brackets :P

**Review by Windchild292 — APPROVED** [review]
Just the odd brackets to fix

**Review by Windchild292 — APPROVED** [review]

**Pull request merged**